securedrop/tests/test_2fa.py

```diff
@@ -1,17 +1,73 @@
 # -*- coding: utf-8 -*-
 import os
+import pytest
+import time
 
+from contextlib import contextmanager
+from datetime import datetime, timedelta
 from flask import url_for
+from pyotp import TOTP
 import flask_testing
 
 os.environ['SECUREDROP_ENV'] = 'test'  # noqa
 from models import Journalist, BadTokenException
+from utils.instrument import InstrumentedApp
 
 import models
 import journalist
 import utils
 
 
+@contextmanager
+def totp_window():
+    # To ensure we have enough time during a single TOTP window to do the
+    # whole test, optionally sleep.
+    now = datetime.now()
+    mod = now.second % 30
+    if mod < 3:
+        time.sleep(mod % 30)
+        now = datetime.now()
+
+    yield
+
+    # This check ensures that the token was used during the same window
+    # in the event that the app's logic only checks for token reuse if the
+    # token was valid.
+    then = datetime.now()
+    assert then < now + timedelta(seconds=(30 - mod))
+
+
+def test_totp_reuse_protections(journalist_app, test_journo):
+    """Ensure that logging in twice with the same TOTP token fails."""
+    original_hardening = models.LOGIN_HARDENING
+    try:
+        models.LOGIN_HARDENING = True
+
+        with totp_window():
+            token = TOTP(test_journo['otp_secret']).now()
+
+            with journalist_app.test_client() as app:
+                with InstrumentedApp(journalist_app) as ins:
+                    resp = app.post('/login',
+                                    data=dict(username=test_journo['username'],
+                                              password=test_journo['password'],
+                                              token=token))
+                    ins.assert_redirects(resp, '/')
+                resp = app.get('/logout', follow_redirects=True)
```

Comment:
I suggest we continue to have a helper `login_user` function, because I think its use makes the code more expressive. What do you think?